### PR TITLE
[Copy update to assert for overlap storage at destination buffer]

### PR DESCRIPTION
**Pull request by jjsjann123**
Assert that there's no multiple written-to to a single memory location, which
caused corrupted output.
Fixed batched matrix trlu logic, which relies on the previous copy behavior to
support tensors with stride 0 at leading dimension.

This fixes the issue proposed at: https://github.com/pytorch/pytorch/issues/23063

### CODE REVIEW

File: test/test_torch.py
Comment (VitalyFedyunin) [inline test/test_torch.py]: super nit: such one lines makes it hard to understand what is supposed to raise an exception here

File: aten/src/ATen/native/LinearAlgebraUtils.h
Comment (VitalyFedyunin) [inline aten/src/ATen/native/LinearAlgebraUtils.h]: Do we ever have situation when is_contiguous tensor doesn't satisfy `tensor.strides() == default_strides_for_size`. As I can see `defaultStrides` and `conmute_contiguous` using exactly same strides formula.

**Comment by VitalyFedyunin** [comment]
cc @pbelevich

File: aten/src/ATen/native/LinearAlgebraUtils.h
Comment (jjsjann123) [reply aten/src/ATen/native/LinearAlgebraUtils.h]: Here we go. :)
```
In [7]: x = torch.randn(2, 10, 2)                                                                                                    

In [8]: y = x[::2, :, :]                                                                                                             

In [9]: y.is_contiguous()                                                                                                            
Out[9]: True

In [10]: y.size()                                                                                                                    
Out[10]: torch.Size([1, 10, 2])

In [11]: y.stride()                                                                                                                  
Out[11]: (40, 2, 1)```

File: test/test_torch.py
Comment (jjsjann123) [reply test/test_torch.py]: Agreed.
Will comment for the test case.

File: aten/src/ATen/native/LinearAlgebraUtils.h
Comment (jjsjann123) [reply aten/src/ATen/native/LinearAlgebraUtils.h]: dimension with size 1 always got me as well

File: aten/src/ATen/native/Copy.cpp
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Copy.cpp]: You can do `iter.check_and_add_output(self)` instead

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by vincentqb** [comment]
@VitalyFedyunin -- I'll let you land this PR from the "needs land" queue.

**Comment by ezyang** [comment]
Waiting on CI after merge.

**Pull request closed without merging**
